unittest/mysys/dynamic_array-t.c

```diff
@@ -0,0 +1,235 @@
+/* Copyright (c) 2024 Eric Herman and MariaDB Foundation.
+
+   This program is free software; you can redistribute it and/or
+   modify it under the terms of the GNU General Public License as
+   published by the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful, but
+   WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1335  USA */
+
+#include <my_global.h>
+#include <my_sys.h>
+#include <tap.h>
+
+struct thing {
+  unsigned id;
+  char name[40];
+};
+
+#define THING_STACK_BUF_SIZE 3
+
+int dyn_array_round_trip(size_t grow_to_size)
+{
+  struct thing things_buf[THING_STACK_BUF_SIZE];
+
+  DYNAMIC_ARRAY things_array;
+  PSI_memory_key psi_key= PSI_NOT_INSTRUMENTED;
+  size_t element_size= sizeof(struct thing);
+  void *init_buffer= (void *)&things_buf;
+  size_t init_alloc= THING_STACK_BUF_SIZE;
+  size_t alloc_increment= 2;
+  myf my_flags= MYF(MY_WME);
+  my_bool err= FALSE;
+
+  ok(1, "THING_STACK_BUF_SIZE: %d, grow_to_size: %zu", THING_STACK_BUF_SIZE,
+     grow_to_size);
+
+  err= my_init_dynamic_array2(psi_key, &things_array, element_size, init_buffer,
+                              init_alloc, alloc_increment, my_flags);
+
+  ok(!err, "my_init_dynamic_array2");
+  if (err) {
+    return EXIT_FAILURE;
+  }
+
+  for (size_t i= 0; i < grow_to_size; ++i) {
+    struct thing tmp_thing;
+
+    tmp_thing.id= i;
+    snprintf(tmp_thing.name, sizeof(tmp_thing.name), "thing %zu", i);
+
+    err= insert_dynamic(&things_array, &tmp_thing);
+    ok(!err, "insert_dynamic for %zu", i);
+    if (err) {
+      delete_dynamic(&things_array);
+      return EXIT_FAILURE;
+    }
+  }
+  ok(grow_to_size == things_array.elements, "size expect: %zu, actual: %zu",
+     grow_to_size, things_array.elements);
+
+  for (size_t i= 0; i < things_array.elements; ++i) {
+    struct thing retrieved;
+    memset(&retrieved, 0x00, sizeof(struct thing));
+
+    get_dynamic(&things_array, &retrieved, i);
+    ok(retrieved.id == i, "thing %zu: id: %u, name: '%s'",
+       i, retrieved.id, retrieved.name);
```

Comment:
trusting it retrieved the name correctly?

Reply:
I view the intent of the test is to check the order.

Do you think there is a test-worthy scenario to guard against where it would have the id, but not the name?
Like perhaps where a struct could be partially copied in to the DYNAMIC_ARRAY?

Reply:
Thanks for addressing review comments to make it a easier to change later.

Noted on order.

Yes, I was considering a bounds check on the contents of the array element copied back/forward. Checking bounds is going to be expanding the contents etc so while trivial, it can be added later.

While I'm not on leave, and I remember the good status of this is, I'm going to merge this now. If you want to expand this to include bounds checking feel free to do so in a new PR under same MDEV.